src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/mirrorpoint/create/actions/EmitUpdateRulesRequestsAction.java

```diff
@@ -0,0 +1,134 @@
+import org.openkilda.wfm.share.model.MirrorContext;
+import org.openkilda.wfm.topology.flowhs.exception.FlowProcessingException;
+import org.openkilda.wfm.topology.flowhs.fsm.common.actions.FlowProcessingAction;
+import org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.create.FlowMirrorPointCreateContext;
+import org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.create.FlowMirrorPointCreateFsm;
+import org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.create.FlowMirrorPointCreateFsm.Event;
+import org.openkilda.wfm.topology.flowhs.fsm.mirrorpoint.create.FlowMirrorPointCreateFsm.State;
+import org.openkilda.wfm.topology.flowhs.model.RequestedFlowMirrorPoint;
+import org.openkilda.wfm.topology.flowhs.service.FlowCommandBuilder;
+import org.openkilda.wfm.topology.flowhs.service.FlowCommandBuilderFactory;
+import org.openkilda.wfm.topology.flowhs.utils.SpeakerInstallSegmentEmitter;
+import org.openkilda.wfm.topology.flowhs.utils.SpeakerRemoveSegmentEmitter;
+import org.openkilda.wfm.topology.flowhs.utils.SpeakerRequestEmitter;
+
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+
+@Slf4j
+public class EmitUpdateRulesRequestsAction
+        extends FlowProcessingAction<FlowMirrorPointCreateFsm, State, Event, FlowMirrorPointCreateContext> {
+    private final FlowCommandBuilderFactory commandBuilderFactory;
+    private final FlowMirrorPointsRepository flowMirrorPointsRepository;
+
+    public EmitUpdateRulesRequestsAction(PersistenceManager persistenceManager,
+                                         FlowResourcesManager resourcesManager) {
+        super(persistenceManager);
+        commandBuilderFactory = new FlowCommandBuilderFactory(resourcesManager);
+        flowMirrorPointsRepository = persistenceManager.getRepositoryFactory().createFlowMirrorPointsRepository();
+    }
+
+    @Override
+    protected void perform(State from, State to,
+                           Event event, FlowMirrorPointCreateContext context, FlowMirrorPointCreateFsm stateMachine) {
+        stateMachine.getCommands().clear();
+        stateMachine.getPendingCommands().clear();
+
+        String flowId = stateMachine.getFlowId();
+        Flow flow = getFlow(flowId);
+
+        FlowCommandBuilder commandBuilder = commandBuilderFactory.getBuilder(flow.getEncapsulationType());
+
+        Collection<FlowSegmentRequestFactory> commands = buildCommands(commandBuilder, stateMachine, flow);
+
+        // emitting
+        PathId flowPathId = stateMachine.getFlowPathId();
+        SwitchId mirrorSwitchId = stateMachine.getMirrorSwitchId();
+        FlowMirrorPoints mirrorPoints = flowMirrorPointsRepository.findByPathIdAndSwitchId(flowPathId, mirrorSwitchId)
+                .orElse(null);
+
+        SpeakerRequestEmitter requestEmitter;
+        if (mirrorPoints != null && mirrorPoints.getMirrorPaths().isEmpty()) {
+            requestEmitter = SpeakerRemoveSegmentEmitter.INSTANCE;
```

Comment:
Still can't get why "create" request emmits remove commands? Create must install segments, remove must remove segments... What I am missing?

Reply:
This action is used both to set new rules and to revert in case of failure.